src/GraphQL/Attributes/ComplexityAttribute.cs

```diff
@@ -0,0 +1,120 @@
+    /// </remarks>
+    public ComplexityAttribute(
+        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)]
+        Type fieldComplexityAnalyzer)
+    {
+        if (!typeof(IFieldComplexityAnalyzer).IsAssignableFrom(fieldComplexityAnalyzer))
+            throw new ArgumentOutOfRangeException(nameof(fieldComplexityAnalyzer), $"The type '{fieldComplexityAnalyzer.GetFriendlyName()}' must implement '{typeof(IFieldComplexityAnalyzer).GetFriendlyName()}'.");
+        if (!fieldComplexityAnalyzer.IsClass)
+            throw new ArgumentOutOfRangeException(nameof(fieldComplexityAnalyzer), $"The type '{fieldComplexityAnalyzer.GetFriendlyName()}' must be a class.");
+        // any other issues, like if value was an open generic or does not have a public parameterless
+        // constructor, will be thrown during the call to CreateInstance
+        _fieldComplexityAnalyzerDelegate = ((IFieldComplexityAnalyzer)Activator.CreateInstance(fieldComplexityAnalyzer)!).Analyze;
+    }
+
+    /// <inheritdoc/>
+    public override void Modify(FieldConfig field)
+        => ModifyField(field);
+
+    /// <inheritdoc/>
+    public override void Modify(FieldType fieldType, bool isInputType)
+    {
+        if (!isInputType)
+            ModifyField(fieldType);
+    }
+
+    private void ModifyField(IFieldMetadataWriter field)
+    {
+        if (_fieldComplexityAnalyzerDelegate != null)
+        {
+            field.WithComplexityImpact(_fieldComplexityAnalyzerDelegate);
+        }
+        else if (FieldImpact != null)
+        {
+            if (ChildImpactMultiplier != null)
+                field.WithComplexityImpact(FieldImpact.Value, ChildImpactMultiplier.Value);
+            else
+                field.WithComplexityImpact(FieldImpact.Value);
+        }
+        else if (ChildImpactMultiplier != null)
+        {
+            field.WithComplexityImpact(context =>
+            {
+                var ret = context.Configuration.DefaultComplexityImpactDelegate(context);
+                ret.ChildImpactMultiplier = ChildImpactMultiplier.Value;
+                return ret;
+            });
+        }
+    }
+}
+
+/// <summary>
+/// Specifies the complexity analyzer for a field.
+/// Not applicable to input fields.
+/// </summary>
+[AttributeUsage(AttributeTargets.Method | AttributeTargets.Field | AttributeTargets.Property, AllowMultiple = false)]
+public class ComplexityAttribute<T> : GraphQLAttribute
+    where T : class, IFieldComplexityAnalyzer, new()
+{
+    private readonly Func<FieldImpactContext, FieldComplexityResult> _func = new T().Analyze;
+
+    /// <inheritdoc/>
+    public override void Modify(FieldConfig field)
+        => field.WithComplexityImpact(_func);
+
+    /// <inheritdoc/>
+    public override void Modify(FieldType fieldType, bool isInputType)
+    {
+        if (!isInputType)
+            fieldType.WithComplexityImpact(_func);
```

Comment:
Silent ignore? Should throw instead? Do we need an analyzer similar to `AllowedOn<>` for code first?

Reply:
Type-first classes may often be used for both input and output types. I would not throw or analyze.